salt/modules/zypper.py

```diff
@@ -2033,3 +2123,85 @@ def list_installed_patches():
         salt '*' pkg.list_installed_patches
     '''
     return _get_patches(installed_only=True)
+
+def list_provides(**kwargs):
+    '''
+    List package provides currently installed as a dict.
+    {'<provided_name>': ['<package_name', 'package_name', ...]}
+    '''
+    if 'pkg.list_provides' in __context__:
+        cached = __context__['pkg.list_provides']
+        return cached
+
+    cmd = ['rpm', '-qa', '--queryformat', '[%{PROVIDES}_|-%{NAME}\n]']
+    ret = {}
+    for line in __salt__['cmd.run'](cmd, output_loglevel='trace', python_shell=False).splitlines():
+        provide, realname = line.split('_|-')
+
+        if provide == realname:
+            continue
+        if provide not in ret:
+            ret[provide] = list()
+        ret[provide].append(realname)
+
+    __context__['pkg.list_list_provides'] = ret
+
+    return ret
```

Comment:
Multiple returns aren't very much welcome. I would do instead:
```python
anchor = 'zypper.list_provides_cache'
ret = __context__.get(anchor)
if not ret:
    ...
    # All that stuff with RPM query
    ...
return ret
```

Reply:
But I still don't like it. How do we invalidate that cache? What happens if that list gets updated? I would maybe capture Zypper's refresh point (it knows better when this is needed) and thus invalidate the data in this case.

Reply:
yep, I will change it.